libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.hpp

```diff
@@ -0,0 +1,93 @@
+/** @file
+ *****************************************************************************
+
+ Declaration of interfaces for top-level SHA256 gadgets.
+
+ *****************************************************************************
+ * @author     This file is part of libsnark, developed by SCIPR Lab
+ *             and contributors (see AUTHORS).
+ * @copyright  MIT license (see LICENSE file)
+ *****************************************************************************/
+
+#ifndef LIBSNARK_GADGETLIB1_GADGETS_HASHES_ANEMOI_COMPONENTS_HPP_
+#define LIBSNARK_GADGETLIB1_GADGETS_HASHES_ANEMOI_COMPONENTS_HPP_
+
+#include <libsnark/common/data_structures/merkle_tree.hpp>
+#include <libsnark/gadgetlib1/gadgets/basic_gadgets.hpp>
+#include <libsnark/gadgetlib1/gadgets/hashes/hash_io.hpp>
+#include <libsnark/gadgetlib1/gadgets/hashes/sha256/sha256_components.hpp>
+
+namespace libsnark
+{
+
+/// Compute y = alpha x^2 + beta
+/// x: input
+/// y: output
+/// alpha, beta: constants
+template<typename FieldT> class anemoi_power_two_gadget : public gadget<FieldT>
+{
+private:
+    // constants
+    FieldT alpha;
+    FieldT beta;
+
+public:
+    // input/output
+    const pb_variable<FieldT> input;
+    const pb_variable<FieldT> output;
+
+    anemoi_power_two_gadget(
+        protoboard<FieldT> &pb,
+        const pb_variable<FieldT> &input,
+        const pb_variable<FieldT> &output,
+        const std::string &annotation_prefix = "")
+        : gadget<FieldT>(pb, annotation_prefix), input(input), output(output)
+    {
+        alpha = FieldT(2);
```

Comment:
these can be assigned in the member initialization list:
```cpp
, alpha(2)
, beta(5)
```

Reply:
Just curious ... are these 2 and 5 for all fields?

Reply:
alpha and beta can be `const`

Reply:
> these can be assigned in the member initialization list:
> 
> ```c++
> , alpha(2)
> , beta(5)
> ```

Done.

Reply:
> alpha and beta can be `const`

Done.

Reply:
> Just curious ... are these 2 and 5 for all fields?

Good question! The $2$ is field/curve-specific and is equal to the generator of the multiplicative subgroup G1. The $5$ is not field-specific, but can have a range of values e.g. $3, 5, 7, 9, \ldots$. The specific value is chosen by the implementer based on a security (higher value) vs. efficieny (lower value) trade-off. All permissible values are listed in Table 1 in the eprint paper https://eprint.iacr.org/2022/840. 

More generally, Anemoi is specified with 4 constants: $\alpha=3,5,7,9, \ldots$, $\beta = G$, $\gamma = 0$, $\delta = G^{-1}$, where $G$ is the generator of the multiplicative subgroup. Prompted by your question, I've created a separate macro for each of constant.